### ISSUE TITLE
Media keys, the neverending saga.

### ISSUE THREAD

**Issue by minusf** [iina/iina#2822]
**System and IINA version:**

- macOS catalina 10.15.3
- IINA 1.0.7-beta1

**Expected behavior:**

Working media keys.

**Actual behavior:**

I simply cannot make the media keys work.  I removed and readdded IINA in Accessibility under Security and Privacy. In settings tried both with and without "Use system media control".  I am using the mpv Default keybindings. Nothing.  `mpv`, which is not a macOS native app can do this.

Would it be possible to have something in IINA that helps debugging this issue?

<details>
<summary><b>Crash report:</b></summary>

```
```
</details>

<br>

<details>
<summary><b>mpv log:</b></summary>

```
```
</details>

<br>

**Steps to reproduce:**

- [x] MPV does not have this problem.

**How often does this happen?**

**Comment by lhc70000** [comment]
The media keys support is based on the system's Media Player framework. If it's working normally, you are supposed to see the title of the current playing video in the "Now Playing" widget in the Today view, and the buttons should also be able to pause/resume/seek the video:

![photo_2020-03-16_21-23-47](https://user-images.githubusercontent.com/8478049/76762555-77419f00-67cc-11ea-93b3-c1d702efc62f.jpg)

If this widget works well, it means that IINA is communicating with the Media Player framework smoothly, and I can't think of any reason at IINA's side to explain your problem. Most stuff is handled by the system automatically.

"Use system media control" should be enabled, and it should have nothing to do with accessibility. If it still doesn't work, maybe you need to:

- Find out whether there's another app hijacking the media keys (lots of legacy apps do this because there's no Media Player framework back then, including IINA v0.0.x);
- Check if iTunes works (Now Playing widget etc). Both iTunes and IINA work on my system, though.
- Restart the system.

**Comment by minusf** [comment]
> Check if iTunes works

there is no iTunes on catalina AFAIK?

This widget is empty atm while i play anything in IINA.

**Comment by minusf** [comment]
All other apps I use that need media keys had to be added to Accesibility under Privacy and Security. This is FAQ with some media players...

**Comment by lhc70000** [comment]
> there is no iTunes on catalina AFAIK?

Sorry, I mean the Music App.

> All other apps I use that need media keys had to be added to Accesibility under Privacy and Security.

Yes, I think these apps are using the legacy way to hijack media keys. They require this permission because they need access to low-level key events. So if they do this way, all other apps won't respond to media keys if any one of them is opened.

I'm not sure whether the Media Player framework is not working, though. As I said, IINA did very little work and the media keys are supposed to be handled by the system……

**Comment by minusf** [comment]
how could i test/make work this "Media Player" framework?

Pre catalina i had `chmod 000 iTunes.app` to stop it hijacking the media keys...

**Comment by lhc70000** [comment]
> how could i test/make work this "Media Player" framework?

It works if the current playing video is shown in the Now Playing widget.

AFAIK the Music App won't hijack the media keys now (it uses the same Media Player framework).

**Comment by minusf** [comment]
No, nothing shows in the widget while playing videos with IINA.  any ideas what i could try?

**Comment by minusf** [comment]
The problem will be on my side as the Music.app itself "does nothing" when i press the media keys. Any ideas to track this down would be most welcome :D

**Comment by minusf** [comment]
this i managed to fix now with `launchctl load -F /System/Library/LaunchAgents/com.apple.rcd.plist`.  IINA now correctly handles the media keys and it shows up in the Playing now widget as well. I disabled this service long time ago to stop iTunes from coming up when pressing F8.  It was not working :}  I ended up doing a `chmod 000` on iTunes after disabling SIP. But now iTunes is gone. and the service remained disabled.

**Comment by minusf** [comment]
I think I'll close this...  Thanks for the help.
